File: Forge/src/main/java/gay/object/hexdebug/forge/HexDebugAbstractionsImpl.java
package gay.object.hexdebug.forge;

import gay.object.hexdebug.IHexDebugAbstractions;
import net.minecraft.server.MinecraftServer;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.event.server.ServerStartedEvent;
import net.minecraftforge.event.server.ServerStoppingEvent;
import net.minecraftforge.fml.loading.FMLPaths;

import java.nio.file.Path;
import java.util.function.Consumer;

public class HexDebugAbstractionsImpl implements IHexDebugAbstractions {
    public static IHexDebugAbstractions get() {
        return new HexDebugAbstractionsImpl();
    }

    @Override
    public Path getConfigDirectory() {
        return FMLPaths.CONFIGDIR.get();
    }

    @Override
    public void initPlatformSpecific() {
        HexDebugConfigForge.init();
    }

    @Override
    public void onServerStarted(Consumer<MinecraftServer> callback) {
        // TODO: is this how you do this???
        MinecraftForge.EVENT_BUS.addListener((Consumer<ServerStartedEvent>) event -> {
            callback.accept(event.getServer());
        });
    }

    @Override
    public void onServerStopping(Consumer<MinecraftServer> callback) {
        MinecraftForge.EVENT_BUS.addListener((Consumer<ServerStoppingEvent>) event -> {
            callback.accept(event.getServer());
        });
    }
}

File: Forge/src/main/kotlin/gay/object/hexdebug/forge/HexDebugAbstractionsImpl.kt
@file:JvmName("HexDebugAbstractionsImpl")

package gay.`object`.hexdebug.forge

import gay.`object`.hexdebug.registry.HexDebugRegistrar
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.packs.resources.PreparableReloadListener
import net.minecraftforge.client.event.RegisterClientReloadListenersEvent
import net.minecraftforge.registries.RegisterEvent
import thedarkcolour.kotlinforforge.forge.MOD_BUS

fun <T : Any> initRegistry(registrar: HexDebugRegistrar<T>) {
    MOD_BUS.addListener { event: RegisterEvent ->
        event.register(registrar.registryKey) { helper ->
            registrar.init(helper::register)
        }
    }
}

fun registerClientResourceReloadListener(id: ResourceLocation, listener: PreparableReloadListener) {
    MOD_BUS.addListener { event: RegisterClientReloadListenersEvent ->
        event.registerReloadListener(listener)
    }
}
